File: sources/easygenerator.Web/Templates/Simple course/App/browserSupport.js
define([], function () {
    var ua = navigator.userAgent.toLowerCase();

    var browserSupport = {
        isSupportedMobile: null,
        isSupportedBrowser: null,
        isMobileDevice: null,
        isIE9: null,

        initialize: initialize
    };
    
    return browserSupport;

    function initialize() {
        browserSupport.isIE9 = isIE9();
        browserSupport.isMobileDevice = isMobileDevice();
        browserSupport.isSupportedMobile= isSupportedMobile();
        browserSupport.isSupportedBrowser = isSupportedBrowser();
    }

    function isIE9() {
        var N = navigator.appName, tem,
            M = ua.match(/(chrome|safari|firefox|msie)\/?\s*([\d\.]+)/i) || [];

        M = M[2] ? [M[1], M[2]] : [N, navigator.appVersion, '-?'];
        if (M && (tem = ua.match(/version\/([\.\d]+)/i)) != null) M[2] = tem[1];

        var browser = M[0].toLowerCase();
        var version = parseInt(M[1], 10);

        return browser == "msie" && version == 9;
    };

    function isMobileDevice() {
        if (ua.indexOf("ipod") != -1 ||
            ua.indexOf("iphone") != -1 ||
            ua.indexOf("ipad") != -1 ||
            ua.indexOf("android") != -1)
            return true;
        return false;
    };

    function isSupportedMobile() {
        if (/iP(hone|od|ad)/.test(navigator.platform)) {
            var v = (navigator.appVersion).match(/OS (\d+)_(\d+)_?(\d+)?/);
            if (parseInt(v[1], 10) >= 6) {
                return true;
            }
        }

        //Android is supported but not chrome, opera or firefox
        if (ua.indexOf("android") != -1 &&
            ua.indexOf("opera") == -1 &&
            ua.indexOf("firefox") == -1) {
            return true;
        }
            
        return false;
    };

    function  isSupportedBrowser() {
        // Opera is not supported
        if (navigator.appName.toLowerCase() == "opera" || navigator.userAgent.indexOf("OPR") != -1)
            return false;

        //IE 9+, Chrome 28+, Firefox 22+, Safari 5+ are supported
        var N = navigator.appName, tem,
            M = ua.match(/(chrome|safari|firefox|msie)\/?\s*([\d\.]+)/i) || [];

        M = M[2] ? [M[1], M[2]] : [N, navigator.appVersion, '-?'];
        if (M && (tem = ua.match(/version\/([\.\d]+)/i)) != null) M[2] = tem[1];

        var browser = M[0].toLowerCase();
        var version = parseInt(M[1], 10);

        if (browser == "chrome" && version >= 28 ||
            browser == "msie" && version >= 9 ||
            browser == "firefox" && version >= 22 ||
            browser == "safari" && version >= 533)
            return true;

        // IE 11 is supported
        if (ua.indexOf('rv:11.0') != -1) {
            return true;
        }

        return false;
    };
});

File: sources/easygenerator.Web/Components/BrowserSupport.cs
using System;
using System.Globalization;
using System.Text.RegularExpressions;
using System.Web;
using System.Web.Configuration;

namespace easygenerator.Web.Components
{
    public static class BrowserSupport
    {
        public static bool IsFullySupportedBrowser(string userAgent, HttpBrowserCapabilities browser)
        {
            if (String.IsNullOrEmpty(userAgent))
            {
                return true;
            }
            var ua = userAgent.ToLower();
            return IsSupportedDesktop(ua, browser) || IsSupportedMobile(ua);
        }

        public static bool IsPartiallySupportedBrowser(string userAgent, HttpBrowserCapabilities browser)
        {
            var ua = userAgent.ToLower();
            var version = "";
            if (!browser.Type.ToUpper().Contains("IE"))
            {
                return true;
            }
            version = Regex.Match(browser.Version, @"[\d]+(\.[\d]+)?").ToString();
            if (version.Length == 0)
            {
                return false;
            }
            return Double.Parse(version, CultureInfo.InvariantCulture) >= 9;
        }

        public static bool IsSupportedDevice(string userAgent)
        {
            if (String.IsNullOrEmpty(userAgent))
            {
                return true;
            }
            var ua = userAgent.ToLower();
            var version = "";
            if (IsIos(ua))
            {
                version = Regex.Match(ua, @" [\d]+_").ToString();
                if (version.Length == 0)
                {
                    return false;
                }
                version = version.Substring(1).Replace("_", "");
                return Int32.Parse(version) >= 6;
            }
            if (!IsAndroid(ua))
            {
                return true;
            }
            version = Regex.Match(ua, @"android [\d]+(\.[\d]+)?").ToString();
            if (version.Length < 9)
            {
                return false;
            }
            version = version.Substring(8);
            return Double.Parse(version, CultureInfo.InvariantCulture) >= 4.4;
        }

        private static bool IsSupportedDesktop(string ua, HttpCapabilitiesBase browser)
        {
            var version = "";
            if (IsAndroid(ua) || IsIos(ua))
            {
                return false;
            }
            if (IsMac(ua))
            {
                return Regex.Match(ua, @"opr|opera|firefox").Length == 0;
            }
            if (ua.Contains("rv:11.0"))
            {
                return true;
            }
            if (ua.Contains("firefox"))
            {
                version = Regex.Match(browser.Version, @"[\d]+(\.[\d]+)?").ToString();
                if (version.Length == 0)
                {
                    return false;
                }
                return Double.Parse(version, CultureInfo.InvariantCulture) >= 22;
            }
            if (ua.Contains("opr"))
            {
                return false;
            }
            if (ua.Contains("phantomjs"))
            {
                return true;
            }
            if (!ua.Contains("chrome"))
            {
                return false;
            }

            version = Regex.Match(browser.Version, @"[\d]+(\.[\d]+)?").ToString();
            if (version.Length == 0)
            {
                return false;
            }
            return Double.Parse(version, CultureInfo.InvariantCulture) >= 28;
        }

        private static bool IsSupportedMobile(string ua)
        {
            var version = "";
            if (IsIos(ua))
            {
                version = Regex.Match(ua, @" [\d]+_").ToString();
                if (version.Length == 0)
                {
                    return false;
                }
                version = version.Substring(1).Replace("_", "");
                return Int32.Parse(version) >= 6 && Regex.Match(ua, @"(opr|opera|firefox|chrome|crios)").Length == 0;
            }
            if (!IsAndroid(ua)) return false;

            version = Regex.Match(ua, @"android [\d]+(\.[\d]+)?").ToString();
            if (version.Length < 9)
            {
                return false;
            }
            version = version.Substring(8);
            return Double.Parse(version, CultureInfo.InvariantCulture) >= 4.4 && Regex.Match(ua, @"(opr|opera|firefox)").Length == 0;
        }

        private static bool IsIos(string ua)
        {
            return Regex.Match(ua, @"(ipad|iphone|ipod)").Length != 0;
        }

        private static bool IsMac(string ua)
        {
            return Regex.Match(ua, @"(macintosh|mac os)").Length != 0;
        }

        private static bool IsAndroid(string ua)
        {
            return ua.Contains("android");
        }

    }
}